conans/model/build_info.py

```diff
@@ -481,57 +482,74 @@ def relocate(el):
     def parsed_requires(self):
         return [r.split("::", 1) if "::" in r else (None, r) for r in self.requires]
 
-    def deduce_locations(self, conanfile):
-        pkg_type = conanfile.package_type
-        if self._exe:   # exe is a new field, it should have the correct location
-            return
-        if self._location or self._link_location:
-            if self._type is None or self._type is PackageType.HEADER:
-                raise ConanException("Incorrect cpp_info defining location without type or header")
-            return
-        if self._type not in [None, PackageType.SHARED, PackageType.STATIC, PackageType.APP]:
-            return
+    def _auto_deduce_locations(self, conanfile):
+
+        def _find_matching(dirs, pattern):
+            lib_found = ""
+            for d in dirs:
+                if lib_found:
+                    break
+                if not os.path.exists(d):
+                    continue
+                # If pattern is an exact match
+                if isinstance(pattern, str):
+                    matches = glob.glob(f"{d}/{pattern}")
+                    if matches:
+                        lib_found = matches[0]
+                    continue
+                # Otherwise, it's a regex
+                files = os.listdir(d)
+                for file_name in files:
+                    full_path = os.path.join(d, file_name)
+                    if not os.path.isfile(full_path):  # Make sure that directories are excluded
+                        continue
+                    elif pattern.match(file_name):
+                        # File name could be a symlink, e.g., mylib.1.0.0.so -> mylib.so
+                        if os.path.islink(full_path):
+                            # Important! os.path.realpath returns the final path of the symlink even if it points
+                            # to another symlink, i.e., libmylib.dylib -> libmylib.1.dylib -> libmylib.1.0.0.dylib
+                            # then os.path.realpath("libmylib.1.0.0.dylib") == "libmylib.dylib"
+                            # Note: os.readlink() returns the path which the symbolic link points to.
+                            real_path = os.path.realpath(full_path)
+                            if pattern.match(os.path.basename(real_path)):
+                                lib_found = real_path
+                        else:
+                            lib_found = full_path
```

Comment:
same comment than above, maybe the code is more clear if we `return` or `break` at these points.